### PR TITLE
CI: Test ml compilation of each commit in a PR in lint job

### PR DESCRIPTION
**Pull request by SkySkimmer**
cc @maximedenes

### CODE REVIEW

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -295,16 +295,12 @@ windows32:
...
-    - apk add bash
-    - dev/lint-repository.sh
+  script: dev/lint-repository.sh
   dependencies: []
-  before_script: []
   variables:
-    # we need an unknown amount of history for per-commit linting
-    GIT_DEPTH: ""
+    GIT_DEPTH: "" # we need an unknown amount of history for per-commit linting
+    OPAM_SWITCH: edge
```
Comment (Zimmi48) [inline .gitlab-ci.yml]: Apparently, you cannot use this without defining `OPAM_VARIANT: "+flambda"` as well. See the CI failure.

**Review by Zimmi48 — APPROVED** [review]

**Review by ejgallego — COMMENTED** [review]
Maybe we should add this workflow to the `build:edge+flambda:dune`, even if the check target is light IMO no point on redoing the build again. Also an early warning may be desirable.

A risk of this feature is that for some infrastructure changes it may prevent stacking too many PRs, but we can live with that IMO, however it should be documented in the CI README, also contributing may need updating, likely with some hints on doing `git rebase -x 'make check'` etc...

**Comment by Zimmi48** [comment]
I agree that `CONTRIBUTING.md` should be updated to expand the description of what the linter is checking.

**Comment by maximedenes** [comment]
@SkySkimmer Thanks for this!

**Comment by ejgallego** [comment]
By the way we had planned to implement this before, but it got on hold until a reliable `dune build @check` was working; the current one is IMHO good enough but we should not be surprised if it fails in very rare occasions [for example moving directories in `theories`]

File: .gitlab-ci.yml [hunk 2]
```diff
@@ -295,16 +295,13 @@ windows32:
...
```
Comment (ejgallego) [inline .gitlab-ci.yml]: IMHO using the `base` compiler could be a better idea here, it is going to be faster than the flambda build very likely; even if in theory they should be the same speed.

**Review by ejgallego — APPROVED** [review]
Looks good to me, I am not very expert on `git` + `bash` stuff.

I think the documentation should suggest users how to do the manual check, this works when doing a rebase:
```
git rebase master --exec 'make -f Makefile.dune check'
```

**Comment by ejgallego** [comment]
> I think the documentation should suggest users how to do the manual check, this works when doing a rebase:

Yeah, add something like that @SkySkimmer and we can merge.

File: .gitlab-ci.yml (hunk 2 above)
Comment (SkySkimmer) [reply .gitlab-ci.yml]: changted

**Comment by SkySkimmer** [comment]
Instead of telling people to use a git command I just told them to use our linter script.

**Comment by ejgallego** [comment]
> Instead of telling people to use a git command I just told them to use our linter script.

Ok, that works for me if you want; on the other hand I think the `git` command is more practical for developers, as it will stop the rebase and allow you to amend.

@Zimmi48 @maximedenes WDTY?

**Comment by SkySkimmer** [comment]
I don't think this is the right place to teach people how to rebase, it 
would just be confusing.

Gaëtan Gilbert

**Comment by ejgallego** [comment]
> I don't think this is the right place to teach people how to rebase, it would just be confusing.

I fear that developers may not know about this rebase functionality, thus leading to bad workflows; I dunno.

Not a blocker to me by far but I'd like to gather some feedback.

**Comment by ejgallego** [comment]
This is in, @coq/contributors / people merging PRs, this is an important change to CI so it makes sense for PRs that have a few commits to be re-run on the new version of the CI.

**Pull request merged**